### PR TITLE
preserve vlen string dtypes, allow vlen string fill_values

### LINKED ISSUE #7652: Saving and loading an array of strings changes datatype to object
### What is your issue?

See the code below
```python
import xarray as xr
# Make an xarray with an array of strings
da = xr.DataArray(
    data=[["a", "b"], ["c", "d"]],
    dims=["x", "y"],
    coords={"x": [0, 1], "y": [0, 1]},
)
da.to_netcdf("test.nc", mode='w')
# Load the xarray back in
da_loaded = xr.load_dataarray("test.nc")
assert da.dtype == da_loaded.dtype, "Dtypes don't match"
```

Now `da_loaded.dtype` is `dtype('O')`.

Same happens with `engine="h5netcdf"`.

```
import xarray
xarray.show_versions()
```
```
INSTALLED VERSIONS
------------------
commit: None
python: 3.11.0 | packaged by conda-forge | (main, Jan 14 2023, 12:26:40) [Clang 14.0.6 ]
python-bits: 64
OS: Darwin
OS-release: 22.3.0
machine: arm64
processor: arm
byteorder: little
LC_ALL: en_US.UTF-8
LANG: en_US.UTF-8
LOCALE: ('en_US', 'UTF-8')
libhdf5: 1.12.2
libnetcdf: None

xarray: 2023.2.0
pandas: 1.5.3
numpy: 1.24.2
scipy: 1.10.0
netCDF4: None
pydap: None
h5netcdf: 1.1.0
h5py: 3.8.0
Nio: None
zarr: None
cftime: None
nc_time_axis: None
PseudoNetCDF: None
rasterio: None
cfgrib: None
iris: None
bottleneck: None
dask: None
distributed: None
matplotlib: 3.7.0
cartopy: None
seaborn: 0.12.2
numbagg: None
fsspec: 2023.1.0
cupy: None
pint: None
sparse: None
flox: None
numpy_groupies: None
setuptools: 67.3.2
pip: 23.0
conda: None
pytest: 7.2.1
mypy: 1.0.1
IPython: 8.10.0
sphinx: 5.3.0
```

### LINKED ISSUE #7868: `open_dataset` with `chunks="auto"` fails when a netCDF4 variables/coordinates is encoded as `NC_STRING`
### What is your issue?

I noticed that `open_dataset` with `chunks="auto"` fails when netCDF4 variables/coordinates are encoded as `NC_STRING`.
The reason is that xarray reads netCDF4 `NC_STRING` as `object` type, and `dask` cannot estimate the size of a `object` dtype. 

As a workaround, the user must currently rewrite the netCDF4 and specify the string DataArray(s) `encoding`(s) as a fixed-length string type (i.e `"S2"` if max string length is 2) so that the data are written as `NC_CHAR` and xarray read it back as byte-encoded fixed-length string type.

**Here below I provide a reproducible example** 

```
import xarray as xr
import numpy as np

# Define string datarray
arr = np.array(["M6", "M3"], dtype=str)
print(arr.dtype)  # <U2
da = xr.DataArray(data=arr, dims=("time"))
data_vars = {"str_arr": da}

# Create dataset
ds_nc_string = xr.Dataset(data_vars=data_vars)

# Set chunking to see behaviour at read-time
ds_nc_string["str_arr"] = ds_nc_string["str_arr"].chunk(1)  # chunks ((1,1),)
 
# Write dataset with NC_STRING
ds_nc_string["str_arr"].encoding["dtype"] = str
ds_nc_string.to_netcdf("/tmp/nc_string.nc")

# Write dataset with NC_CHAR
ds_nc_char = xr.Dataset(data_vars=data_vars)
ds_nc_char["str_arr"].encoding["dtype"] = "S2"
ds_nc_char.to_netcdf("/tmp/nc_char.nc")

# When NC_STRING, chunks="auto" does not work when string are saved as 
# --> NC STRING is read as object, and dask can not estimate chunk size !
# If chunks={} it reads the NC_STRING array in a single dask chunk !!!
ds_nc_string = xr.open_dataset("/tmp/nc_string.nc", chunks="auto") # NotImplementedError
ds_nc_string = xr.open_dataset("/tmp/nc_string.nc", chunks={})     # Works
ds_nc_string.chunks  # chunks (2,)

# With NC_CHAR, chunks={} and chunks="auto" works and returns the same result!
ds_nc_char = xr.open_dataset("/tmp/nc_char.nc", chunks={})   
ds_nc_char.chunks # chunks (2,)
ds_nc_char = xr.open_dataset("/tmp/nc_char.nc", chunks="auto")
ds_nc_char.chunks # chunks (2,)

# NC_STRING is read back as object 
ds_nc_string = xr.open_dataset("/tmp/nc_string.nc", chunks=None)
ds_nc_string["str_arr"].dtype  #  object 

# NC_CHAR is read back as fixed length byte-string representation (S2) 
ds_nc_char = xr.open_dataset("/tmp/nc_char.nc", chunks=None)
ds_nc_char["str_arr"].dtype            #  S2 
ds_nc_char["str_arr"].data.astype(str) #  U2 
```

Questions: 
- `open_dataset` should not take care of automatically deserializing the `NC_CHAR` fixed-length byte-string representation into a `Unicode string`? 
- `open_dataset` should not take care of automatically reading `NC_STRING` as `Unicode string` (converting `object` to `str`)? 

Related issues are: 
- https://github.com/pydata/xarray/issues/7652
- https://github.com/pydata/xarray/issues/2059
- https://github.com/pydata/xarray/pull/7654
- https://github.com/pydata/xarray/issues/2040

### PR DESCRIPTION
**Pull request by kmuehlbauer**
- [x] Closes (partly) #1647
- [x] Closes #7652 
- [x] Closes #7868
- [x] Tests adapted
- [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [ ] New functions/methods are listed in `api.rst`

### CODE REVIEW

**Comment by kmuehlbauer** [comment]
Resurrected this PR, rebased on latest main and force-pushed.

**Comment by kmuehlbauer** [comment]
So, the h5netcdf-part does only work for versions h5py >= 3.7 and h5netcdf >= 1.1.0 (this fails for the 3.9 min deps). How can I account for that?  Is there some best-practice?

**Comment by kmuehlbauer** [comment]
> So, the h5netcdf-part does only work for versions h5py >= 3.7 and h5netcdf >= 1.1.0 (this fails for the 3.9 min deps). How can I account for that? Is there some best-practice?

We are already unblocked for the h5py-version as the minimum dependency version is already at 3.7. h5netcdf get's unblocked on November, 23rd (in two month's time).

Should we just wait it out and make vlen strings available in one block?

**Comment by dcherian** [comment]
> h5netcdf get's unblocked on November, 23rd

I'm OK with bumping it early for this. If you want we could wait till after the next release to bring it down to one month :)

**Comment by kmuehlbauer** [comment]
> I'm OK with bumping it early for this. If you want we could wait till after the next release to bring it down to one month :)

Yes, wait until after next release and bump early.

File: ci/requirements/min-all-deps.yml [hunk 1]
```diff
@@ -16,11 +16,11 @@ dependencies:
   - dask-core=2022.7
   - distributed=2022.7
   - flox=0.5
-  - h5netcdf=1.0
+  - h5netcdf=1.1
   # h5py and hdf5 tend to cause conflicts
   # for e.g. hdf5 1.12 conflicts with h5py=3.1
   # prioritize bumping other packages instead
-  - h5py=3.6
+  - h5py=3.7
```
Comment (dcherian) [inline ci/requirements/min-all-deps.yml:23]: can we add this to whats-new please

File: xarray/coding/variables.py [hunk 2]
```diff
@@ -562,3 +562,15 @@ def encode(self, variable: Variable, name: T_Name = None) -> Variable:
 
     def decode(self):
         raise NotImplementedError()
+
+
+class ObjectStringCoder(VariableCoder):
```
Comment (dcherian) [inline xarray/coding/variables.py:567]:
```suggestion
class ObjectVLenStringCoder(VariableCoder):
```
Maybe?

Comment (kmuehlbauer) [reply xarray/coding/variables.py:567]: Yes, more explicit. Thanks!

**Comment by kmuehlbauer** [comment]
@dcherian Finally, I think this is ready.

**Comment by dcherian** [comment]
Thanks @kmuehlbauer

**Pull request merged**